Unable to build version 0.8 for AWS Lambda

I tried several ways to get ifcopenshell to work with AWS Lambda. For reference, I am using EFS to host my Python packages, which are later imported in my lambda function:

```
sys.path.append('/mnt/efs')
import ifcopenshell
import ifcopenshell.util
```

However, the only version that I am able to import in Lambda successfully is the one that is mentioned in the Dockerfile:

`IFC_OPENSHELL_BUILD="39-v0.7.0-476ab50"`

This is the updated Dockerfile that I made to get v0.8:

```
# Dockerfile for Running AWS Lambda Function with Python and IfcOpenShell

# Base image: Python 3.9 from AWS's public container registry
FROM public.ecr.aws/docker/library/python:3.12 AS build

# Set the location of your Lambda function code
ARG FUNCTION_DIR="/var/task"

# Install necessary packages
RUN apt-get -y update && apt-get -y install unzip wget

# Install AWS Lambda runtime interface client
RUN pip install --target ${FUNCTION_DIR} awslambdaric

# Download and extract IfcOpenShell
RUN wget https://github.com/IfcOpenShell/IfcOpenShell/releases/download/ifcopenshell-python-0.8.0/ifcopenshell-python-0.8.0-py312-linux64.zip && \
    unzip ifcopenshell-python-0.8.0-py312-linux64.zip && \
    mv ifcopenshell ${FUNCTION_DIR} && \
    rm ifcopenshell-python-0.8.0-py312-linux64.zip

# Copy the requirements file and install dependencies
COPY requirements.txt .
RUN pip install -r requirements.txt

# Copy the Lambda function code
COPY ./example_handler ${FUNCTION_DIR}/example_handler

# Set the working directory
WORKDIR ${FUNCTION_DIR}

# Set the entrypoint for the Lambda function
ENTRYPOINT [ "/usr/local/bin/python", "-m", "awslambdaric" ]

# Set the Python import path to the Lambda function handler
# This path is relative to the root of the Lambda function code (FUNCTION_DIR)
# Lambda invocation will look for this path
CMD ["example_handler.extract_wall_psets_handler"]
```
But when I copy this ifcopenshell folder to my imports, I get this error:

`IfcOpenShell not built for 'linux/64bit/python3.12'`

For v0.8, I have tried many different packages, both from the official releases:
https://docs.ifcopenshell.org/ifcopenshell-python/installation.html#zip-packages

And also the wheels:
https://pypi.org/project/ifcopenshell/0.8.0/#files

I am not able to get it to work for any Python version. 

Is there something that I'm missing? How can I get v0.8 to work with my Lambda function?

Thank you for your reply!

I ran the dockerfile that you gave and copied the ifcopenshell folder to my EFS storage, but the lambda function gives the same error. This is what I did:

```
sudo docker build -t lambda_python12_v8 .

[+] Building 0.3s (11/11) FINISHED                                                                           docker:default
 => [internal] load build definition from Dockerfile                                                                   0.0s
 => => transferring dockerfile: 1.03kB                                                                                 0.0s
 => [internal] load metadata for public.ecr.aws/docker/library/python:3.12                                             0.2s
 => [internal] load .dockerignore                                                                                      0.0s
 => => transferring context: 2B                                                                                        0.0s
 => [1/7] FROM public.ecr.aws/docker/library/python:3.12@sha256:251ef8e69b6ccdf3c7bf7effaa51179d59af35364dd9c86469142  0.0s
 => CACHED [2/7] RUN apt-get -y update && apt-get -y install unzip wget                                                0.0s
 => CACHED [3/7] RUN pip install --target /var/task awslambdaric                                                       0.0s
 => CACHED [4/7] RUN wget https://github.com/IfcOpenShell/IfcOpenShell/releases/download/ifcopenshell-python-0.8.0/if  0.0s
 => CACHED [5/7] WORKDIR /var/task                                                                                     0.0s
 => CACHED [6/7] RUN /usr/local/bin/python -c "import ifcopenshell; f = ifcopenshell.file(); f.write('/test.ifc')"     0.0s
 => CACHED [7/7] RUN cat /test.ifc                                                                                     0.0s
 => exporting to image                                                                                                 0.0s
 => => exporting layers                                                                                                0.0s
 => => writing image sha256:13af56b29a0621511dda980a251bc916954a414f4f83e9c0bb421c38364e1028                           0.0s
 => => naming to docker.io/library/lambda_python12_v8                                                                  0.0s
```

```
sudo docker run --name lambda_python12_v8_container lambda_python12_v8
```

```
sudo docker cp lambda_python12_v8_container:/var/task ./p12v8

Successfully copied 181MB to /home/ec2-user/IfcOpenShell/aws/lambda/p12v8
```

```
sudo cp -r p12v8/ifcopenshell/ /efs/efs
```

And this is my test lambda function:

```
import os
import sys
sys.path.append('/mnt/efs')

print(os.listdir('/mnt/efs'))

import ifcopenshell
import ifcopenshell.util


def lambda_handler(event, context):
    print(ifcopenshell.version)
```

Output:

```
Status: Failed
Test Event Name: test

Response:
{
  "errorMessage": "Unable to import module 'lambda_function': IfcOpenShell not built for 'linux/64bit/python3.12'",
  "errorType": "Runtime.ImportModuleError",
  "requestId": "",
  "stackTrace": []
}

Function Logs:
['ifcopenshell']
[ERROR] Runtime.ImportModuleError: Unable to import module 'lambda_function': IfcOpenShell not built for 'linux/64bit/python3.12'
Traceback (most recent call last):
INIT_REPORT Init Duration: 391.75 ms	Phase: init	Status: error	Error Type: Runtime.ImportModuleError
['ifcopenshell']
[ERROR] Runtime.ImportModuleError: Unable to import module 'lambda_function': IfcOpenShell not built for 'linux/64bit/python3.12'
Traceback (most recent call last):
INIT_REPORT Init Duration: 2894.25 ms	Phase: invoke	Status: error	Error Type: Runtime.ImportModuleError
```

--- a/src/ifcopenshell-python/ifcopenshell/__init__.py
+++ b/src/ifcopenshell-python/ifcopenshell/__init__.py
@@ -85,8 +85,19 @@ sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "lib", p
 
 try:
     from . import ifcopenshell_wrapper
-except Exception:
-    raise ImportError("IfcOpenShell not built for '%s'" % python_distribution)
+except Exception as e:
+    import importlib.machinery
+
+    if any(
+        os.path.exists(os.path.join(directory, "_ifcopenshell_wrapper" + suffix))
+        for directory in (
+            os.path.dirname(__file__),
+            os.path.join(os.path.dirname(__file__), "lib", python_distribution),
+        )
+        for suffix in importlib.machinery.EXTENSION_SUFFIXES
+    ):
+        raise ImportError("IfcOpenShell for '%s' failed to load: %s" % (python_distribution, e)) from e
+    raise ImportError("IfcOpenShell not built for '%s'" % python_distribution) from e
 
 # `_file`, `_stream` is used only for annotations inside this file,
 # see https://github.com/microsoft/pyright/discussions/9065.

--- a/src/ifcopenshell-python/ifcopenshell/__init__.py
+++ b/src/ifcopenshell-python/ifcopenshell/__init__.py
@@ -86,18 +86,7 @@ sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "lib", p
 try:
     from . import ifcopenshell_wrapper
 except Exception as e:
-    import importlib.machinery
-
-    if any(
-        os.path.exists(os.path.join(directory, "_ifcopenshell_wrapper" + suffix))
-        for directory in (
-            os.path.dirname(__file__),
-            os.path.join(os.path.dirname(__file__), "lib", python_distribution),
-        )
-        for suffix in importlib.machinery.EXTENSION_SUFFIXES
-    ):
-        raise ImportError("IfcOpenShell for '%s' failed to load: %s" % (python_distribution, e)) from e
-    raise ImportError("IfcOpenShell not built for '%s'" % python_distribution) from e
+    raise ImportError("IfcOpenShell not built for '%s' (%s)" % (python_distribution, e)) from e
 
 # `_file`, `_stream` is used only for annotations inside this file,
 # see https://github.com/microsoft/pyright/discussions/9065.

